File: src/rendering/Polygon.js
class Polygon {
  /*
   * @param {[Point]} vertices
   * @param {string} edge_color - The color of the edges of the polygon in the format of a CSS color
   * value (see: https://developer.mozilla.org/en-US/docs/Web/CSS/color_value). RGBA format is
   * "rgb(#,#,#,#)"
   * @param {string} fill_color - The color of the inside of the polygon in the format of a CSS color
   * value (see: https://developer.mozilla.org/en-US/docs/Web/CSS/color_value). RGBA format is
   * "rgba(#,#,#,#)"
   */
  constructor(vertices, edge_color, fill_color) {
    this.vertices = vertices;
    this.edge_color = edge_color;
    this.fill_color = fill_color;
  }

  /*
   * @param {Object} polygon_json - A JSON object to create a Polygon from.
   * @returns {Polygon} - A Polygon object that was created from the specified JSON object.
   */
  fromJson(polygon_json) {
    this.vertices = polygon_json.vertices.map(point_json => (new Point()).fromJson(point_json));
    let ec = polygon_json.edge_color;
    this.edge_color = 'rgba(' + ec.r + ',' + ec.g + ',' + ec.b + ',' + ec.a + ')';
    let fc = polygon_json.fill_color;
    this.fill_color = 'rgba(' + fc.r + ',' + fc.g + ',' + fc.b + ',' + fc.a + ')';

    return this;
  }

  /*
   * @brief - Renders the polygon.
   * @param {CanvasRenderingContext2D} context - The rendering context to render the polygon to.
   */
  render(context) {
    context.strokeStyle = this.edge_color;
    context.fillStyle = this.fill_color;

    context.beginPath();
    let firstVertex = this.vertices[0];
    context.moveTo(firstVertex.x, firstVertex.y);
    for (let vertex_idx = 1; vertex_idx < this.vertices.length; ++vertex_idx) {
      let currVertex = this.vertices[vertex_idx];
      context.lineTo(currVertex.x, currVertex.y);
    }
    context.closePath();

    context.stroke();
    context.fill();
  }
}

File: src/rendering/polygon.cpp
#include "rendering/polygon.hpp"

namespace cpp_web_viz {

Polygon::Polygon()
  : vertices({}), edge_color({0, 0, 0, 0}), fill_color({0, 0, 0, 0}) {}

Polygon::Polygon(const std::vector<PositionInPixels>& vertices_, const RGBA& edge_color_,
                 const RGBA& fill_color_)
    : vertices(vertices_), edge_color(edge_color_), fill_color(fill_color_) {}

void Polygon::Translate(const PositionInPixels& t) {
  std::transform(vertices.begin(), vertices.end(), vertices.begin(),
    [&t](const PositionInPixels& vertex) -> PositionInPixels {
      return vertex + t;
    });
}

void to_json(nlohmann::json& json, const Polygon& polygon) {
  json = nlohmann::json{
    {"vertices", polygon.vertices},
    {"edge_color", polygon.edge_color},
    {"fill_color", polygon.fill_color}};
}

}  // namespace cpp_web_viz
